pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go

```diff
@@ -123,23 +119,76 @@ func (d *stateData) updateClaimStatus(ctx context.Context, clientset kubernetes.
 	return nil
 }
 
-// initializePodSchedulingContext can be called concurrently. It returns an existing PodSchedulingContext
-// object if there is one already, retrieves one if not, or as a last resort creates
-// one from scratch.
-func (d *stateData) initializePodSchedulingContexts(ctx context.Context, pod *v1.Pod, podSchedulingContextLister resourcev1alpha2listers.PodSchedulingContextLister) (*resourcev1alpha2.PodSchedulingContext, error) {
-	// TODO (#113701): check if this mutex locking can be avoided by calling initializePodSchedulingContext during PreFilter.
-	d.mutex.Lock()
-	defer d.mutex.Unlock()
+type podSchedulingState struct {
+	// A pointer to the PodSchedulingContext object for the pod, if one exists
+	// in the API server.
+	//
+	// Conceptually, this object belongs into the scheduler framework
+	// where it might get shared by different plugins. But in practice,
+	// it is currently only used by dynamic provisioning and thus
+	// managed entirely here.
+	schedulingCtx *resourcev1alpha2.PodSchedulingContext
+
+	// selectedNode is set if (and only if) a node has been selected.
+	selectedNode *string
 
-	if d.schedulingCtx != nil {
-		return d.schedulingCtx, nil
-	}
+	// potentialNodes is set if (and only if) the potential nodes field
+	// needs to be updated or set.
+	potentialNodes *[]string
```

Comment:
is a pointer to a nil/empty slice possible? Meaning that we would clear all potentialNodes?

Reply:
Yes, that could happen and is supported by making this a pointer. Otherwise it would be impossible to distinguish "clear the field" from "leave it as-is".